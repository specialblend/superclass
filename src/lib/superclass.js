import assert from 'assert';
import isConstructable from '@specialblend/is-constructable';
import { copyPrototype } from './common';

/**
 * create a defaultExport from a parent class and provided sister classes
 * @param {Class} base: the base class to extend
 * @param {Array<Class>} supertypes: the sister classes
 * @returns {Class}: the created defaultExport
 */
export const superclass = (base, ...supertypes) => {
    assert(isConstructable(base), 'base must be constructable');
    const subtype = class extends base {
        constructor(...props) {
            super(...props);
            const originalObj = Object.assign({}, this);
            const superprops = {};
            for (const Supertype of supertypes.reverse()) {
                Object.assign(superprops, new Supertype(...props));
            }
            Object.assign(this, superprops, originalObj);
        }
    };
    for (const supertype of supertypes.reverse()) {
        copyPrototype(subtype, supertype);
    }
    copyPrototype(subtype, base);
    return subtype;
};
