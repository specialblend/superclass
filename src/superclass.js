import * as R from 'ramda';
import { PrettyError } from '@specialblend/pretty-error';
import util from 'util';

const copyPrototype = (target, source) => {
    for (const prop of Reflect.ownKeys(source.prototype)) {
        if (typeof prop === 'symbol' || !prop.match(/^(?:constructor|prototype|arguments|caller|name|bind|call|apply|toString|length)$/)) {
            if (typeof target.prototype[prop] === 'undefined') {
                Object.assign(target.prototype, { [prop]: source.prototype[prop] });
            }
        }
    }
};

const matches = R.flip(R.ifElse(R.is(Function), R.is, R.equals));

/**
 * Superclass Symbol, used to identity Superclass
 * @type {symbol}
 * @private
 */
export const __superclass__ = Symbol('__superclass__');

export class SuperclassError extends PrettyError {}

/**
 * Assert provided object is equal to or instance of allowed types
 * @param provided
 * @param name
 * @param allowed
 * @returns {boolean}
 */
const assertType = (provided, name, ...allowed) => {
    if (R.any(matches(provided), allowed)) {
        return true;
    }
    throw new SuperclassError(`provided invalid type for ${name}:`, {
        allowed,
        provided,
    });
};

/**
 * create a subclass of provided base class whose constructor
 * will receive arguments transformed by provided transformer function
 * @param {Class} base: the base class to extend
 * @param {Function} transform: the transformer function
 * @returns {Class}: the created subclass
 */
export const mixin = (base, transform) => {
    assertType(base, `argument 1 of ${util.inspect(mixin)}`, Function);
    assertType(transform, `argument 2 of ${util.inspect(mixin)}`, Function, null, undefined);
    return class extends base {
        constructor(...props) {
            if (transform === null) {
                super();
                return;
            }
            if (typeof transform === 'undefined') {
                super(...props);
                return;
            }
            if (typeof transform === 'function') {
                const transformedProps = transform(...props);
                assertType(transformedProps, `result of calling ${util.inspect(transform)}`, Array);
                super(...transformedProps);
            }
        }
    };
};

/**
 * create a Superclass from a parent class and provided sister classes
 * @param {Class} base: the base class to extend
 * @param {Array<Class>} supertypes: the sister classes
 * @returns {Class}: the created Superclass
 */
export const superclass = (base, ...supertypes) => {
    assertType(base, `argument 1 of ${util.inspect(superclass)}`, Function);
    const subtype = class extends base {
        constructor(...props) {
            super(...props);
            for (const Supertype of supertypes.reverse()) {
                const obj = {};
                Object.assign(obj, new Supertype(...props), this);
                Object.assign(this, obj);
            }
        }
    };
    for (const supertype of supertypes) {
        copyPrototype(subtype, supertype);
    }
    return Object.assign(subtype, { [__superclass__]: __superclass__ });
};
