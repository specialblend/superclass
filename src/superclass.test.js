/* eslint-disable new-cap */

import { mixin, superclass, SuperclassError } from './superclass';

describe('mixin', () => {
    test('is a function', () => {
        expect(mixin).toBeFunction();
    });
    test('throws expected error on invalid base', () => {
        try {
            mixin('this-is;not/a~valid_class');
        } catch (err) {
            expect(err).toBeInstanceOf(SuperclassError);
            expect(err.message).toMatchSnapshot();
        }
    });
    test('throws expected error on invalid transformer', () => {
        try {
            mixin(Object, 'this-is;not/a~valid_transformer');
        } catch (err) {
            expect(err).toBeInstanceOf(SuperclassError);
            expect(err.message).toMatchSnapshot();
        }
    });
    test('throws expected error on invalid transformer result', () => {
        try {
            new class extends mixin(Object, () => null) {};
        } catch (err) {
            expect(err).toBeInstanceOf(SuperclassError);
            expect(err.message).toMatchSnapshot();
        }
    });
    describe('works as expected', () => {
        test('returns a function', () => {
            expect(mixin(Object)).toBeFunction();
        });
        test('without transformer', () => {
            class Hello {
                constructor(msg) {
                    this.message = msg;
                }
            }
            class HelloToo extends mixin(Hello) {}
            const testMessage = 'test.hello.message1234567890-';
            const hello = new Hello(testMessage);
            const helloToo = new HelloToo(testMessage);
            expect(hello).toHaveProperty('message', testMessage);
            expect(helloToo).toHaveProperty('message', testMessage);
        });
        test('with null transformer', () => {
            class Hello {
                constructor(msg) {
                    this.message = msg;
                }
            }
            class FuckYou extends mixin(Hello, null) {}
            const testMessage = 'test.hello.message1234567890-';
            const hello = new Hello(testMessage);
            const helloToo = new FuckYou(testMessage);
            expect(hello).toHaveProperty('message', testMessage);
            expect(helloToo).toHaveProperty('message', undefined);
        });
        test('with function transformer', () => {
            class Hello {
                constructor(msg) {
                    this.message = msg;
                }
            }
            class ExcitedHello extends mixin(Hello, msg => [msg.toUpperCase()]) {}
            const testMessage = 'test.hello.message1234567890-';
            const hello = new Hello(testMessage);
            const helloToo = new ExcitedHello(testMessage);
            expect(hello).toHaveProperty('message', testMessage);
            expect(helloToo).toHaveProperty('message', testMessage.toUpperCase());
        });
    });
});

describe('superclass', () => {
    const __foo__ = Symbol('__foo__');
    const __bar__ = Symbol('__bar__');
    const __baz__ = Symbol('__baz__');
    const __faz__ = Symbol('__faz__');
    const __superFoo__ = Symbol('__superFoo__');
    const __superDuperFoo__ = Symbol('__superDuperFoo__');

    class SaysHello {
        constructor(message) {
            this.message = message;
            if (typeof this.messages === 'undefined') {
                this.messages = {};
            }
            this.messages[this.constructor.name] = message;
            this[this.constructor.name] = () => this.constructor.name;
        }
        hello() {
            const {
                message,
                messages: { [this.constructor.name]: namedMessage },
            } = this;
            return {
                message,
                namedMessage,
            };
        }
    }

    class Foo extends SaysHello {
        [__foo__]() {
            return __foo__;
        }
    }

    class Bar extends SaysHello {
        [__foo__]() {
            return __bar__;
        }
        [__bar__]() {
            return __bar__;
        }
    }

    class Baz extends SaysHello {
        [__foo__]() {
            return __bar__;
        }
        [__bar__]() {
            return __bar__;
        }
        [__baz__]() {
            return __baz__;
        }
    }

    class SuperFoo extends superclass(Foo) {
        [__superFoo__]() {

        }
    }

    class SuperDuperFoo extends superclass(SuperFoo) {
        [__superDuperFoo__]() {

        }
    }

    class Faz extends superclass(Foo, Bar, Baz) {
        [__faz__]() {

        }
    }

    const message = 'They say it dont be like it is, but it do!';
    const superFoo = new SuperFoo(message);
    const superDuperFoo = new SuperDuperFoo(message);
    const faz = new Faz(message);
    test('is a function', () => {
        expect(superclass).toBeFunction();
    });
    test('throws expected error on invalid base', () => {
        try {
            superclass('this-is;not/a~valid_class');
        } catch (err) {
            expect(err).toBeInstanceOf(SuperclassError);
            expect(err.message).toMatchSnapshot();
        }
    });
    describe('of single Superclass', () => {
        test('returns a function', () => {
            expect(SuperFoo).toBeFunction();
        });
        describe('instance', () => {
            test('has expected own props', () => {
                expect(typeof superFoo[__superFoo__]).toBe('function');
                expect(superFoo).toHaveProperty('message', message);
                expect(superFoo).toHaveProperty('messages', expect.any(Object));
                expect(superFoo.messages).toHaveProperty('SuperFoo', message);
                expect(superFoo).toHaveProperty('SuperFoo', expect.any(Function));
                expect(superFoo.SuperFoo()).toBe('SuperFoo');
                expect(superFoo).toHaveProperty('hello', expect.any(Function));
                expect(superFoo.hello()).toMatchObject({
                    message,
                    namedMessage: message,
                });
            });
            test('has expected super props', () => {
                expect(typeof superFoo[__foo__]).toBe('function');
            });
        });
        describe('of single Superclass', () => {
            test('returns a function', () => {
                expect(SuperDuperFoo).toBeFunction();
            });
            describe('instance', () => {
                test('has expected props', () => {
                    expect(typeof superDuperFoo[__superDuperFoo__]).toBe('function');
                    expect(superDuperFoo).toHaveProperty('message', message);
                    expect(superDuperFoo).toHaveProperty('messages', expect.any(Object));
                    expect(superDuperFoo.messages).toHaveProperty('SuperDuperFoo', message);
                    expect(superDuperFoo).toHaveProperty('SuperDuperFoo', expect.any(Function));
                    expect(superDuperFoo.SuperDuperFoo()).toBe('SuperDuperFoo');
                    expect(superDuperFoo).toHaveProperty('hello', expect.any(Function));
                    expect(superDuperFoo.hello()).toMatchObject({
                        message,
                        namedMessage: message,
                    });
                });
                test('has expected super props', () => {
                    expect(typeof superDuperFoo[__superFoo__]).toBe('function');
                    expect(typeof superDuperFoo[__foo__]).toBe('function');
                });
            });
        });
    });
    describe('of multiple Superclasses', () => {
        test('returns a function', () => {
            expect(Faz).toBeFunction();
        });
        describe('instance', () => {
            test('has expected own props', () => {
                expect(typeof faz[__faz__]).toBe('function');
                expect(faz).toHaveProperty('message', message);
                expect(faz).toHaveProperty('messages', expect.any(Object));
                expect(faz.messages).toHaveProperty('Faz', message);
                expect(faz).toHaveProperty('Faz', expect.any(Function));
                expect(faz.Faz()).toBe('Faz');
                expect(faz).toHaveProperty('hello', expect.any(Function));
                expect(faz.hello()).toMatchObject({
                    message,
                    namedMessage: message,
                });
            });
            test('has expected super props', () => {
                expect(typeof faz[__foo__]).toBe('function');
                expect(typeof faz[__bar__]).toBe('function');
                expect(typeof faz[__baz__]).toBe('function');
            });
            test('respects order of operations', () => {
                expect(faz[__foo__]()).toBe(__foo__);
                expect(faz[__bar__]()).toBe(__bar__);
                expect(faz[__baz__]()).toBe(__baz__);
            });
        });
    });
    describe('works as expected with mixins', () => {
        const say = jest.fn();
        class SayHello {
            constructor(name) {
                say(`Hello ${name}`);
            }
        }
        class SayWhatsUp {
            constructor(name) {
                say(`Whats up ${name}`);
            }
        }
        class SayBye {
            constructor(name) {
                say(`Bye ${name}`);
            }
        }
        const ExcitedSayHello = mixin(SayHello, name => [name.toUpperCase()]);
        const BoredWhatsUp = mixin(SayWhatsUp, name => [`${name}...`]);
        const HatefulSayBye = mixin(SayBye, name => [`${name}. See you never!`]);
        class BipolarPerson extends superclass(ExcitedSayHello, BoredWhatsUp, HatefulSayBye) {}
        test('calls say with expected data', () => {
            new BipolarPerson('John Smith');
            expect(say).toHaveBeenCalledWith('Hello JOHN SMITH');
            expect(say).toHaveBeenCalledWith('Whats up John Smith...');
            expect(say).toHaveBeenCalledWith('Bye John Smith. See you never!');
        });
    });
});
