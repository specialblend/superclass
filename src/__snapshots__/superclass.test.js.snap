// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mixin throws expected error on invalid base 1`] = `
"provided invalid type for argument 1 of [Function: mixin]:
{
  allowed: [
    [Function: Function]
  ],
  provided: 'this-is;not/a~valid_class'
}"
`;

exports[`mixin throws expected error on invalid base and transformer 1`] = `
"provided invalid type for argument 1 of [Function: mixin]:
{
  allowed: [
    [Function: Function]
  ],
  provided: 'this-is;not/a~valid_class'
}"
`;

exports[`mixin throws expected error on invalid base and valid function transformer 1`] = `
"provided invalid type for argument 1 of [Function: mixin]:
{
  allowed: [
    [Function: Function]
  ],
  provided: 'this-is;not/a~valid_class'
}"
`;

exports[`mixin throws expected error on invalid transformer 1`] = `
"provided invalid type for argument 2 of [Function: mixin]:
{
  allowed: [
    [Function: Function],
    null,
    undefined
  ],
  provided: 'this-is;not/a~valid_transformer'
}"
`;

exports[`mixin throws expected error on invalid transformer result 1`] = `
"provided invalid type for result of calling [Function]:
{
  allowed: [
    [Function: Array]
  ],
  provided: null
}"
`;

exports[`superclass throws expected error on invalid base 1`] = `
"provided invalid type for argument 1 of [Function: superclass]:
{
  allowed: [
    [Function: Function]
  ],
  provided: 'this-is;not/a~valid_class'
}"
`;
